File: mobilosignal/app/src/main/java/barilyuk/mobilosignal/BootReceiver.java
package barilyuk.mobilosignal;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.util.Log;
import android.os.Build;

public class BootReceiver extends BroadcastReceiver {
    private static final String TAG = "BootReceiver";
    private static final String PREFS_NAME = "SIMSelection";
    private static final String AUTO_START_KEY = "AutoStart";

    @Override
    public void onReceive(Context context, Intent intent) {
        Log.d(TAG, "onReceive: Received intent: " + intent.getAction());

        if (Intent.ACTION_BOOT_COMPLETED.equals(intent.getAction())) {
            SharedPreferences prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE);
            boolean isAutoStartEnabled = prefs.getBoolean(AUTO_START_KEY, false);

            Log.d(TAG, "onReceive: Auto start enabled: " + isAutoStartEnabled);

            if (isAutoStartEnabled) {
                Log.d(TAG, "onReceive: Auto start enabled, starting SignalStrengthService");
                Intent serviceIntent = new Intent(context, SignalStrengthService.class);

                try {
                    // For devices running Android O (API level 26) or higher, use startForegroundService
                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                        context.startForegroundService(serviceIntent);
                    } else {
                        context.startService(serviceIntent);
                    }
                    Log.d(TAG, "onReceive: Service start attempted successfully");
                } catch (Exception e) {
                    Log.e(TAG, "onReceive: Failed to start service", e);
                }
            } else {
                Log.d(TAG, "onReceive: Auto start not enabled");
            }
        }
    }
}

File: mobilosignal/app/src/main/java/barilyuk/mobilosignal/BootReceiver.kt
package barilyuk.mobilosignal

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.util.Log

class BootReceiver : BroadcastReceiver() {

    override fun onReceive(context: Context, intent: Intent) {
        if (intent.action != Intent.ACTION_BOOT_COMPLETED) return
        if (!Prefs.isAutoStartEnabled(context)) {
            Log.d(TAG, "Auto start disabled, ignoring boot")
            return
        }

        Log.d(TAG, "Auto start enabled, starting SignalStrengthService")
        runCatching {
            context.startForegroundService(Intent(context, SignalStrengthService::class.java))
        }.onFailure { Log.e(TAG, "Failed to start service on boot", it) }
    }

    private companion object {
        const val TAG = "BootReceiver"
    }
}
